### DISCUSSION TITLE
Add support for non-auto-increment default-value based primary keys (like UUID)

### DISCUSSION THREAD

**Discussion by calsch03** [directus/directus#5472] [Feature Requests]
I have connected Directus with an existing MS SQL database. The primary keys use the datatype uniqueidentifier and this results in an error when I change the field in Directus or use it in a M2O relationship.

1) The issue and what you expected to happen
I can create new items using the `newid()` command in the primary key field. But when I edit the primary key field in the Directus data model Directus gives an invalid payload error.
![image](https://user-images.githubusercontent.com/19907926/116527925-7f2ea680-a8db-11eb-82d0-7e9b373dce7c.png)

2) Exact steps to reproduce this issue
- Create a MS SQL database and a table with a primary key of data type uniqueidentifier.
- Connect Directus to the MS SQL database.
- In Directus data model: change the primary key field (for example make it read only) and Directus throws the error.

3) Your environment:
- Directus: 9.0.0-rc.61
- MS SQL server: 15.0.2080.9
- Which deployment are you using: npx
- What browser are you using: Microsoft Edge 90.0.818.46

**Comment by rijkvanzanten** [comment]
The same applies to Postgres' `generate_uuid()` (plugin) function, and potentially other default value based values

**Comment by oreilles** [comment]
~~I added support for `uniqueidentifier` there https://github.com/directus/directus/pull/5804.~~ it's merged
If your `uniqueidentifier` column has a default value set to `NEWID()`, it still won't work, because Directus will actually try to insert the string `"NEWID()"` (see discussion in https://github.com/directus/directus/pull/5699)
But now that Directus sees your field as a `uuid`, you'll be able to configure the `On Create` option to `Generate new UUID` within the app - Directus will now generate a valid `uuid` on creation and send it in the payload, which will work.

**Reply by calsch03** [reply]
Good work on supporting the `uniqueidentifier` data type. I was able to insert new items to tables with `NEWID()` as the default value with no problems. I recently switched to PostgreSQL and also use the `uuid_generate_v4()` as default value here. But using the build-in Directus function for this is generally the better solution I agree.

**Reply by oreilles** [reply]
> I was able to insert new items to tables with NEWID() as the default value with no problems.

Wow you're right, it works.
It probably shouldn't though because it means that the default value sent to the database was unescaped. Is this intended @rijkvanzanten  ?

**Reply by rijkvanzanten** [reply]
We currently rely on Knex to do the sanitation of values, so what you enter is passed on to [`.defaultTo()`](http://knexjs.org/#Schema-defaultTo) as is, yeah. 🤔

**Reply by oreilles** [reply]
@calsch03 How did you make it work with PostgreSQL ? Did you set the default value to `uuid_generate_v4()` in Directus or in the database ?

**Reply by rijkvanzanten** [reply]
@Oreilles If Directus is generating the UUID still, it never reaches the default value which might explain why it works?

**Reply by oreilles** [reply]
> If Directus is generating the UUID still, it never reaches the default value which might explain why it works?

Oh, that's what was happening. I had left the option to generate it automatically. @calsch03 must have had set the default in the database, not in Directus.

**Reply by calsch03** [reply]
> @calsch03 How did you make it work with PostgreSQL ? Did you set the default value to `uuid_generate_v4()` in Directus or in the database ?

Yeah I use `uuid_generate_v4()` as default in the database tables for primary keys. Directus automatically assigned it as default value after connecting it to the database.

**Reply by oreilles** [reply]
The database does it, not Directus 😉 Directus can't use functions as default values, or any values at all because everything sent to the database is escaped by knex to prevent SQL injection.

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
